packages/stateManager/index.ts

```diff
@@ -36,67 +77,293 @@ abstract class PublicObjectManager<
     super();
   }
   abstract _processEvent(event: schema.ShopEvents): Promise<boolean>;
-  abstract create(obj: T): Promise<[any, any]>;
-  abstract update(obj: T): Promise<[any, any]>;
-  get(id: string): Promise<T> {
-    return this.db.get(id);
-  }
+  abstract get(key: string): Promise<T>;
+  abstract getAll(): T[];
 }
 
-class ItemManager extends PublicObjectManager<Item> {
+class ListingManager extends PublicObjectManager<Item> {
   constructor(db: Store<Item>, client: RelayClient) {
     super(db, client);
   }
   // Process the events; if the event modifies one the shops items this updates the store and emits an event
   // returns a bool to indicate whether the event was processed or not
   async _processEvent(event: schema.ShopEvents): Promise<boolean> {
-    if (typeof event === schema.ShopEvents.UpdateItem) {
-      // load the item from the store
-      const item = await this.db.get(event.eventId);
-      Object.assign(item, event);
-      await this.db.put(event.eventId, item);
-      this.emit("update", item);
+    if (event.createItem) {
+      const ci = event.createItem;
+      const metadata = JSON.parse(new TextDecoder().decode(ci.metadata));
+      const id = bytesToHex(ci.eventId);
+      const item = {
+        price: ci.price,
+        metadata,
+        tags: [],
+        quantity: 0,
+      };
+      await this.db.put(id, item);
+      return true;
+    } else if (event.updateItem) {
+      const ui = event.updateItem;
+      const id = bytesToHex(ui.itemId);
+      const item = await this.db.get(id);
+      if (ui.metadata) {
+        item.metadata = JSON.parse(new TextDecoder().decode(ui.metadata));
+      }
+      if (ui.price) {
+        item.price = ui.price;
+      }
+      await this.db.put(id, item);
+      return true;
+    } else if (event.changeStock) {
+      const cs = event.changeStock;
+      if (cs.itemIds) {
+        cs.itemIds.map(async (id: Uint8Array, i: number) => {
+          const itemId = bytesToHex(id);
+          const item = await this.db.get(itemId);
+          const diff = cs.diffs ? cs.diffs[i] : 0;
+          item.quantity = item.quantity + diff;
+          await this.db.put(itemId, item);
+        });
+        // if the changeStock event is emitted from a payment(AKA has txHash), it has to also go through OrderStore
+        if (cs.txHash) {
+          return false;
+        } else return true;
+      }
+    } else if (event.updateTag) {
```

Comment:
we need these tag events in the product sub level since we are adding and removing tags from a product